README
NAME
    Catalyst::View::Email - Send Email from Catalyst

SYNOPSIS
    This module sends out emails from a stash key specified in the
    configuration settings.

CONFIGURATION
    WARNING: since version 0.10 the configuration options slightly changed!

    Use the helper to create your View:

        $ script/myapp_create.pl view Email Email

    In your app configuration:

        __PACKAGE__->config(
            'View::Email' => {
                # Where to look in the stash for the email information.
                # 'email' is the default, so you don't have to specify it.
                stash_key => 'email',
                # Define the defaults for the mail
                default => {
                    # Defines the default content type (mime type). Mandatory
                    content_type => 'text/plain',
                    # Defines the default charset for every MIME part with the 
                    # content type text.
                    # According to RFC2049 a MIME part without a charset should
                    # be treated as US-ASCII by the mail client.
                    # If the charset is not set it won't be set for all MIME parts
                    # without an overridden one.
                    # Default: none
                    charset => 'utf-8'
                },
                # Setup how to send the email
                # all those options are passed directly to Email::Sender::Simple
                sender => {
                    # if mailer doesn't start with Email::Sender::Simple::Transport::,
                    # then this is prepended.
                    mailer => 'SMTP',
                    # mailer_args is passed directly into Email::Sender::Simple 
                    mailer_args => {
                        host     => 'smtp.example.com', # defaults to localhost
                        username => 'username',
                        password => 'password',
                }
              }
            }
        );

NOTE ON SMTP
    If you use SMTP and don't specify host, it will default to localhost and
    attempt delivery. This often means an email will sit in a queue and not
    be delivered.

SENDING EMAIL
    Sending email is just filling the stash and forwarding to the view:

        sub controller : Private {
            my ( $self, $c ) = @_;

            $c->stash->{email} = {
                to      => 'jshirley@gmail.com',
                cc      => 'abraxxa@cpan.org',
                bcc     => join ',', qw/hidden@secret.com hidden2@foobar.com/,
                from    => 'no-reply@foobar.com',
                subject => 'I am a Catalyst generated email',
                body    => 'Body Body Body',
            };
        
            $c->forward( $c->view('Email') );
        }

    Alternatively you can use a more raw interface and specify the headers
    as an array reference like it is passed to Email::MIME::Creator. Note
    that you may also mix both syntaxes if you like ours better but need to
    specify additional header attributes. The attributes are appended to the
    header array reference without overwriting contained ones.

        $c->stash->{email} = {
            header => [
                To      => 'jshirley@gmail.com',
                Cc      => 'abraxxa@cpan.org',
                Bcc     => join ',', qw/hidden@secret.com hidden2@foobar.com/,
                From    => 'no-reply@foobar.com',
                Subject => 'Note the capitalization differences',
            ],
            body => qq{Ain't got no body, and nobody cares.},
            # Or, send parts
            parts => [
                Email::MIME->create(
                    attributes => {
                        content_type => 'text/plain',
                        disposition  => 'attachment',
                        charset      => 'US-ASCII',
                    },
                    body => qq{Got a body, but didn't get ahead.},
                )
            ],
        };

HANDLING ERRORS
    If the email fails to send, the view will die (throw an exception).
    After your forward to the view, it is a good idea to check for errors:

        $c->forward( $c->view('Email') );
    
        if ( scalar( @{ $c->error } ) ) {
            $c->error(0); # Reset the error condition if you need to
            $c->response->body('Oh noes!');
        } else {
            $c->response->body('Email sent A-OK! (At least as far as we can tell)');
        }

USING TEMPLATES FOR EMAIL
    Now, it's no fun to just send out email using plain strings. Take a look
    at Catalyst::View::Email::Template to see how you can use your favourite
    template engine to render the mail body.

METHODS
    new Validates the base config and creates the Email::Sender::Simple
        object for later use by process.

    process($c)
        The process method does the actual processing when the view is
        dispatched to.

        This method sets up the email parts and hands off to
        Email::Sender::Simple to handle the actual email delivery.

    setup_attributes($c, $attr)
        Merge attributes with the configured defaults. You can override this
        method to return a structure to pass into generate_message which
        subsequently passes the return value of this method to
        Email::MIME->create under the "attributes" key.

    generate_message($c, $attr)
        Generate a message part, which should be an Email::MIME object and
        return it.

        Takes the attributes, merges with the defaults as necessary and
        returns a message object.

TROUBLESHOOTING
    As with most things computer related, things break. Email even more so.
    Typically any errors are going to come from using SMTP as your sending
    method, which means that if you are having trouble the first place to
    look is at Email::Sender::Transport::SMTP. This module is just a wrapper
    for Email::Sender::Simple, so if you get an error on sending, it is
    likely from there anyway.

    If you are using SMTP and have troubles sending, whether it is
    authentication or a very bland "Can't send" message, make sure that you
    have Net::SMTP and, if applicable, Net::SMTP::SSL installed.

    It is very simple to check that you can connect via Net::SMTP, and if
    you do have sending errors the first thing to do is to write a simple
    script that attempts to connect. If it works, it is probably something
    in your configuration so double check there. If it doesn't, well, keep
    modifying the script and/or your mail server configuration until it
    does!

SEE ALSO
  Catalyst::View::Email::Template - Send fancy template emails with Cat
  Catalyst::Manual - The Catalyst Manual
  Catalyst::Manual::Cookbook - The Catalyst Cookbook
AUTHORS
    J. Shirley <jshirley@gmail.com>

    Alexander Hartmaier <abraxxa@cpan.org>

CONTRIBUTORS
    (Thanks!)

    Matt S Trout

    Daniel Westermann-Clark

    Simon Elliott <cpan@browsing.co.uk>

    Roman Filippov

    Lance Brown <lance@bearcircle.net>

    Devin Austin <dhoss@cpan.org>

    Chris Nehren <apeiron@cpan.org>

COPYRIGHT
    Copyright (c) 2007 - 2009 the Catalyst::View::Email "AUTHORS" and
    "CONTRIBUTORS" as listed above.

LICENSE
    This library is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Catalyst::View::Email.

0.30    2010-05-22
        - Added Test::Requires to 06config.t

0.29    2010-05-21
        - Using Test::Requires for Mason and Template::Toolkit dependency check
        - Removed Template::Toolkit and Mason as prereqs

0.28    2010-05-14
        - Added  version numbers to Helpers

0.27    2010-03-23
        - made ::View::TT and ::View::Mason hard deps
        - adjusted tests to reflect the aforementioned
        - patches welcome

0.25_02 2010-02-21
        - cleaned up tests
        - made Test::More 0.88 a prereq

0.25_01 2010-02-21
        - updated/corrected docs
        - added proper svn repo url

0.25    2010-02-01
        - converted to Module::Install

0.23    2010-01-30
        - removed Email::Send::Test from t/lib/* views

0.22    2010-01-28
        - small fixes, version bumpage etc

0.21_01 2010-01-27
        - dev release to test recent fixes

0.21    2010-01-27
        - provided a sane MANIFEST.SKIP for dzil

0.20    2010-01-26
        - fix transport instantiation thanks to Chris Nehren <apeiron@cpan.org>

0.19    2010-01-18
        - fixed optional dependencies
		- added META stuff back in

0.18    2010-01-17
        - Fixed version dependencies for failing regressions
		- added Email::Send::Test as a dependency

0.17    2010-01-14
        - fixed MANIFEST

0.16    2010-01-12
        - Fixed a Makefile.PL issue

0.15    2010-01-12
        - Fixed teeny version issue with Email::Sender::Simple

0.14    2010-01-09
        - Ported over to using Email::Sender::Simple, and Moose
		- Fixed Mason test so that it doesn't choke, as around 'process' seems to make
		  it die.

0.13    2009-02-22 14:06:00
        - bcc: POD corrections thanks to Lance Brown <lance@bearcircle.net>

0.12    2009-01-22 06:52:00
        - Fixing tests for new versions of MIME::Creator
        - Better structure of the code so that ::Template can also handle
          plain text views
        - Added onto troubleshooting

0.11    2008-07-04 09:14:00
        - Fixing a bug where content-type was ignored so multipart/alternative
          failed.  RT #32215

0.10    2007-11-22 23:00:00
        - Refactored by Alexander Hartmaier with api changes
          and POD improvements

0.06
        - Fixing some slight issues with configuration not being handled
          appropriately (thanks dwc and mst)
0.05
        - Better support for configuration
        - Adding Mason tests

0.04
        - Fixing MANIFEST, distribution

0.03
        - Improved handling of configuration, thanks to mst.

0.02
        - Allow configuration from application or ConfigLoader
        - Minor POD updates

0.01  2007-06-03
        - Initial release


t/lib/TestApp.pm
package # Hide from PAUSE
    TestApp;

use Catalyst qw/ -Debug /;
use FindBin;

TestApp->config(
    root            => "$FindBin::Bin/root",
    default_view    => 'TT',
    'View::Email::AppConfig' => {
        sender => {
            mailer => 'Test',
        },
    },
    'View::Email::Template::AppConfig' => {
        stash_key => 'template_email',
        sender => {
            mailer => 'Test',
        },
		content_type => 'text/html',
        default => {
            view => 'TT',
        },
    },
);

TestApp->setup;

1;


t/lib/TestApp/View/TT.pm
package # Hide me.
    TestApp::View::TT;

use strict;
use base 'Catalyst::View::TT';

1;


t/lib/TestApp/View/Email.pm
package # Hide from PAUSE
    TestApp::View::Email;

use Email::Send::Test;

use base 'Catalyst::View::Email';

__PACKAGE__->config(
    sender => {
        mailer => 'Test'
    },
);

1;


Makefile.PL
use inc::Module::Install 0.91;
test_requires 'Test::Requires';

name 'Catalyst-View-Email';
all_from 'lib/Catalyst/View/Email.pm';
author 'J. Shirley <jshirley@gmail.com>';
license 'perl';

test_requires 'Test::More';

requires 'MIME::Base64'          => '3.08';
requires 'Authen::SASL'          => '2.13';
requires 'Catalyst'              => '5.7';
requires 'Moose'                 => '0.93';
requires 'parent'                => '0.223';
requires 'Email::Sender::Simple' => '0.100110';
requires 'Email::MIME'           => '1.859';
requires 'Email::MIME::Creator'  => '1.455';
requires 'Test::More'            => '0.88';
requires 'Catalyst::View::TT'    => '0.31';
#requires 'Catalyst::View::Mason' => '0.18';

# Rebuild README for maintainers
if ($Module::Install::AUTHOR) {
    system("pod2text lib/Catalyst/View/Email.pm > README");
}

realclean_files 'README';

resources 'IRC'     => 'irc://irc.perl.org/#catalyst';
resources 'license' => 'http://dev.perl.org/licenses/';
resources 'repository' =>
  'http://dev.catalyst.perl.org/repos/Catalyst/Catalyst-View-Email/';
resources 'MailingList' =>
  'http://lists.scsys.co.uk/cgi-bin/mailman/listinfo/catalyst';

auto_provides;
auto_install;
WriteAll;


lib/Catalyst/View/Email/Template.pm
package Catalyst::View::Email::Template;

use Moose;
use Carp;
use Scalar::Util qw/ blessed /;
extends 'Catalyst::View::Email';

our $VERSION = '0.29';
$VERSION = eval $VERSION;
=head1 NAME

Catalyst::View::Email::Template - Send Templated Email from Catalyst

=head1 SYNOPSIS

Sends templated mail, based upon your default view. It captures the output
of the rendering path, slurps in based on mime-types and assembles a multi-part
email using L<Email::MIME::Creator> and sends it out.

=head1 CONFIGURATION

WARNING: since version 0.10 the configuration options slightly changed!

Use the helper to create your view:
    
    $ script/myapp_create.pl view Email::Template Email::Template

For basic configuration look at L<Catalyst::View::Email/CONFIGURATION>.

In your app configuration (example in L<YAML>):

    View::Email::Template:
        # Optional prefix to look somewhere under the existing configured
        # template  paths.
        # Default: none
        template_prefix: email
        # Define the defaults for the mail
        default:
            # Defines the default view used to render the templates.
            # If none is specified neither here nor in the stash
            # Catalysts default view is used.
            # Warning: if you don't tell Catalyst explicit which of your views should
            # be its default one, C::V::Email::Template may choose the wrong one!
            view: TT

=head1 SENDING EMAIL

Sending email works just like for L<Catalyst::View::Email> but by specifying 
the template instead of the body and forwarding to your Email::Template view:

    sub controller : Private {
        my ( $self, $c ) = @_;

        $c->stash->{email} = {
            to          => 'jshirley@gmail.com',
            cc          => 'abraxxa@cpan.org',
            bcc         => 'hidden@secret.com hidden2@foobar.com',
            from        => 'no-reply@foobar.com',
            subject     => 'I am a Catalyst generated email',
            template    => 'test.tt',
            content_type => 'multipart/alternative'
        };
        
        $c->forward( $c->view('Email::Template') );
    }

Alternatively if you want more control over your templates you can use the following idiom
to override the defaults:

    templates => [
        {
            template        => 'email/test.html.tt',
            content_type    => 'text/html',
            charset         => 'utf-8',
            view            => 'TT', 
        },
        {
            template        => 'email/test.plain.mason',
            content_type    => 'text/plain',
            charset         => 'utf-8',
            view            => 'Mason', 
        }
    ]


=head1 HANDLING ERRORS

See L<Catalyst::View::Email/HANDLING ERRORS>.

=cut

# here the defaults of Catalyst::View::Email are extended by the additional
# ones Template.pm needs.

has 'stash_key' => (
    is      => 'rw',
    isa     => 'Str',
    default => sub { "email" },
    lazy    => 1,
);

has 'template_prefix' => (
    is      => 'rw',
    isa     => 'Str',
    default => sub { '' },
    lazy    => 1,
);

has 'default' => (
    is      => 'rw',
    isa     => 'HashRef',
    default => sub {
        {
            view         => 'TT',
            content_type => 'text/html',
        };
    },
    lazy => 1,
);

# This view hitches into your default view and will call the render function
# on the templates provided.  This means that you have a layer of abstraction
# and you aren't required to modify your templates based on your desired engine
# (Template Toolkit or Mason, for example).  As long as the view adequately
# supports ->render, all things are good.  Mason, and others, are not good.

#
# The path here is to check configuration for the template root, and then
# proceed to call render on the subsequent templates and stuff each one
# into an Email::MIME container.  The mime-type will be stupidly guessed with
# the subdir on the template.
#

# Set it up so if you have multiple parts, they're alternatives.
# This is on the top-level message, not the individual parts.
#multipart/alternative

sub _validate_view {
    my ( $self, $view ) = @_;

    croak "C::V::Email::Template's configured view '$view' isn't an object!"
      unless ( blessed($view) );

    croak
      "C::V::Email::Template's configured view '$view' isn't an Catalyst::View!"
      unless ( $view->isa('Catalyst::View') );

    croak
"C::V::Email::Template's configured view '$view' doesn't have a render method!"
      unless ( $view->can('render') );
}

=head1 METHODS

=over 4

=item generate_part

Generates a MIME part to include in the email. Since the email is template based
every template piece is a separate part that is included in the email.

=cut

sub generate_part {
    my ( $self, $c, $attrs ) = @_;

    my $template_prefix      = $self->template_prefix;
    my $default_view         = $self->default->{view};
    my $default_content_type = $self->default->{content_type};
    my $default_charset      = $self->default->{charset};

    my $view;

    # use the view specified for the email part
    if (   exists $attrs->{view}
        && defined $attrs->{view}
        && $attrs->{view} ne '' )
    {
        $view = $c->view( $attrs->{view} );
        $c->log->debug(
            "C::V::Email::Template uses specified view $view for rendering.")
          if $c->debug;
    }

    # if none specified use the configured default view
    elsif ($default_view) {
        $view = $c->view($default_view);
        $c->log->debug(
            "C::V::Email::Template uses default view $view for rendering.")
          if $c->debug;
    }

    # else fallback to Catalysts default view
    else {
        $view = $c->view;
        $c->log->debug(
"C::V::Email::Template uses Catalysts default view $view for rendering."
        ) if $c->debug;
    }

    # validate the per template view
    $self->_validate_view($view);

    # prefix with template_prefix if configured
    my $template =
      $template_prefix ne ''
      ? join( '/', $template_prefix, $attrs->{template} )
      : $attrs->{template};

    # setup the attributes (merge with defaults)
    my $e_m_attrs = $self->SUPER::setup_attributes( $c, $attrs );

    # render the email part
    my $output = $view->render(
        $c,
        $template,
        {
            content_type => $e_m_attrs->{content_type},
            stash_key    => $self->stash_key,
            %{$c->stash},
        }
    );
    if ( ref $output ) {
        croak $output->can('as_string') ? $output->as_string : $output;
    }

    return Email::MIME->create(
        attributes => $e_m_attrs,
        body       => $output,
    );
}

=item process

The process method is called when the view is dispatched to. This creates the
multipart message and then sends the message contents off to
L<Catalyst::View::Email> for processing, which in turn hands off to
L<Email::Sender::Simple>.

=cut

around 'process' => sub {
    my ( $orig, $self, $c, @args ) = @_;
    my $stash_key = $self->stash_key;
    return $self->$orig( $c, @args )
      unless $c->stash->{$stash_key}->{template}
          or $c->stash->{$stash_key}->{templates};

    # in case of the simple api only one
    my @parts = ();

    # now find out if the single or multipart api was used
    # prefer the multipart one

    # multipart api
    if (   $c->stash->{$stash_key}->{templates}
        && ref $c->stash->{$stash_key}->{templates} eq 'ARRAY'
        && ref $c->stash->{$stash_key}->{templates}[0] eq 'HASH' )
    {

        # loop through all parts of the mail
        foreach my $part ( @{ $c->stash->{$stash_key}->{templates} } ) {
            push @parts,
              $self->generate_part(
                $c,
                {
                    view         => $part->{view},
                    template     => $part->{template},
                    content_type => $part->{content_type},
                    charset      => $part->{charset},
                }
              );
        }
    }

    # single part api
    elsif ( $c->stash->{$stash_key}->{template} ) {
        push @parts,
          $self->generate_part( $c,
            { template => $c->stash->{$stash_key}->{template}, } );
    }

    delete $c->stash->{$stash_key}->{body};
    $c->stash->{$stash_key}->{parts} ||= [];
    push @{ $c->stash->{$stash_key}->{parts} }, @parts;
    
    return $self->$orig($c);

};

=back

=head1 TODO

=head2 ATTACHMENTS

There needs to be a method to support attachments.  What I am thinking is
something along these lines:

    attachments => [
        # Set the body to a file handle object, specify content_type and
        # the file name. (name is what it is sent at, not the file)
        { body => $fh, name => "foo.pdf", content_type => "application/pdf" },
        # Or, specify a filename that is added, and hey, encoding!
        { filename => "foo.gif", name => "foo.gif", content_type => "application/pdf", encoding => "quoted-printable" },
        # Or, just a path to a file, and do some guesswork for the content type
        "/path/to/somefile.pdf",
    ]

=head1 SEE ALSO

=head2 L<Catalyst::View::Email> - Send plain boring emails with Catalyst

=head2 L<Catalyst::Manual> - The Catalyst Manual

=head2 L<Catalyst::Manual::Cookbook> - The Catalyst Cookbook

=head1 AUTHORS

J. Shirley <jshirley@gmail.com>

Simon Elliott <cpan@browsing.co.uk>

Alexander Hartmaier <abraxxa@cpan.org>

=head1 LICENSE

This library is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut

1;


lib/Catalyst/View/Email.pm
package Catalyst::View::Email;

use Moose;
use Carp;

use Encode qw(encode decode);
use Email::Sender::Simple qw/ sendmail /;
use Email::MIME::Creator;
extends 'Catalyst::View';

our $VERSION = '0.29';
$VERSION = eval $VERSION;

has 'mailer' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { "sendmail" }
);

has '_mailer_obj' => (
    is      => 'rw',
    isa     => 'Email::Sender::Transport',
    lazy    => 1,
    builder => '_build_mailer_obj',
);

has 'stash_key' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { "email" }
);

has 'default' => (
    is      => 'rw',
    isa     => 'HashRef',
    default => sub { { content_type => 'text/plain' } },
    lazy    => 1,
);

has 'sender' => (
    is      => 'rw',
    isa     => 'HashRef',
    lazy    => 1,
    default => sub { { mailer => shift->mailer } }
);

has 'content_type' => (
    is      => 'rw',
    isa     => 'Str',
    default => sub { shift->default->{content_type} },
    lazy    => 1,
);

=head1 NAME

Catalyst::View::Email - Send Email from Catalyst

=head1 SYNOPSIS

This module sends out emails from a stash key specified in the
configuration settings.

=head1 CONFIGURATION

WARNING: since version 0.10 the configuration options slightly changed!

Use the helper to create your View:
    
    $ script/myapp_create.pl view Email Email

In your app configuration:

    __PACKAGE__->config(
        'View::Email' => {
            # Where to look in the stash for the email information.
            # 'email' is the default, so you don't have to specify it.
            stash_key => 'email',
            # Define the defaults for the mail
            default => {
                # Defines the default content type (mime type). Mandatory
                content_type => 'text/plain',
                # Defines the default charset for every MIME part with the 
                # content type text.
                # According to RFC2049 a MIME part without a charset should
                # be treated as US-ASCII by the mail client.
                # If the charset is not set it won't be set for all MIME parts
                # without an overridden one.
                # Default: none
                charset => 'utf-8'
            },
            # Setup how to send the email
            # all those options are passed directly to Email::Sender::Simple
            sender => {
                # if mailer doesn't start with Email::Sender::Simple::Transport::,
                # then this is prepended.
                mailer => 'SMTP',
                # mailer_args is passed directly into Email::Sender::Simple 
                mailer_args => {
                    Host     => 'smtp.example.com', # defaults to localhost
                    username => 'username',
                    password => 'password',
            }
          }
        }
    );

=head1 NOTE ON SMTP

If you use SMTP and don't specify host, it will default to localhost and
attempt delivery. This often means an email will sit in a queue and
not be delivered.

=cut

=head1 SENDING EMAIL

Sending email is just filling the stash and forwarding to the view:

    sub controller : Private {
        my ( $self, $c ) = @_;

        $c->stash->{email} = {
            to      => 'jshirley@gmail.com',
            cc      => 'abraxxa@cpan.org',
            bcc     => join ',', qw/hidden@secret.com hidden2@foobar.com/,
            from    => 'no-reply@foobar.com',
            subject => 'I am a Catalyst generated email',
            body    => 'Body Body Body',
        };
        
        $c->forward( $c->view('Email') );
    }

Alternatively you can use a more raw interface and specify the headers as
an array reference like it is passed to L<Email::MIME::Creator>.
Note that you may also mix both syntaxes if you like ours better but need to
specify additional header attributes.
The attributes are appended to the header array reference without overwriting
contained ones.

    $c->stash->{email} = {
        header => [
            To      => 'jshirley@gmail.com',
            Cc      => 'abraxxa@cpan.org',
            Bcc     => join ',', qw/hidden@secret.com hidden2@foobar.com/,
            From    => 'no-reply@foobar.com',
            Subject => 'Note the capitalization differences',
        ],
        body => qq{Ain't got no body, and nobody cares.},
        # Or, send parts
        parts => [
            Email::MIME->create(
                attributes => {
                    content_type => 'text/plain',
                    disposition  => 'attachment',
                    charset      => 'US-ASCII',
                },
                body => qq{Got a body, but didn't get ahead.},
            )
        ],
    };

=head1 HANDLING ERRORS

If the email fails to send, the view will die (throw an exception).
After your forward to the view, it is a good idea to check for errors:
    
    $c->forward( $c->view('Email') );
    
    if ( scalar( @{ $c->error } ) ) {
        $c->error(0); # Reset the error condition if you need to
        $c->response->body('Oh noes!');
    } else {
        $c->response->body('Email sent A-OK! (At least as far as we can tell)');
    }

=head1 USING TEMPLATES FOR EMAIL

Now, it's no fun to just send out email using plain strings.
Take a look at L<Catalyst::View::Email::Template> to see how you can use your
favourite template engine to render the mail body.

=head1 METHODS

=over 4

=item new

Validates the base config and creates the L<Email::Sender::Simple> object for later use
by process.

=cut

sub BUILD {
    my $self = shift;

    my $stash_key = $self->stash_key;
    croak "$self stash_key isn't defined!"
      if ( $stash_key eq '' );

}

sub _build_mailer_obj {
    my ($self) = @_;
    my $transport_class = ucfirst $self->sender->{mailer};

    # borrowed from Email::Sender::Simple -- apeiron, 2010-01-26
    if ( $transport_class !~ /^Email::Sender::Transport::/ ) {
        $transport_class = "Email::Sender::Transport::$transport_class";
    }

    Class::MOP::load_class($transport_class);

    return $transport_class->new( $self->sender->{mailer_args} || {} );
}

=item process($c)

The process method does the actual processing when the view is dispatched to.

This method sets up the email parts and hands off to L<Email::Sender::Simple> to handle
the actual email delivery.

=cut

sub process {
    my ( $self, $c ) = @_;

    croak "Unable to send mail, bad mail configuration"
      unless $self->sender->{mailer};

    my $email = $c->stash->{ $self->stash_key };
    croak "Can't send email without a valid email structure"
      unless $email;

    # Default content type
    if ( $self->content_type and not $email->{content_type} ) {
        $email->{content_type} = $self->content_type;
    }

    my $header = $email->{header} || [];
    push @$header, ( 'To' => delete $email->{to} )
      if $email->{to};
    push @$header, ( 'Cc' => delete $email->{cc} )
      if $email->{cc};
    push @$header, ( 'Bcc' => delete $email->{bcc} )
      if $email->{bcc};
    push @$header, ( 'From' => delete $email->{from} )
      if $email->{from};
    push @$header,
      ( 'Subject' => Encode::encode( 'MIME-Header', delete $email->{subject} ) )
      if $email->{subject};
    push @$header, ( 'Content-type' => $email->{content_type} )
      if $email->{content_type};

    my $parts = $email->{parts};
    my $body  = $email->{body};

    unless ( $parts or $body ) {
        croak "Can't send email without parts or body, check stash";
    }

    my %mime = ( header => $header, attributes => {} );

    if ( $parts and ref $parts eq 'ARRAY' ) {
        $mime{parts} = $parts;
    }
    else {
        $mime{body} = $body;
    }

    $mime{attributes}->{content_type} = $email->{content_type}
      if $email->{content_type};
    if (    $mime{attributes}
        and not $mime{attributes}->{charset}
        and $self->{default}->{charset} )
    {
        $mime{attributes}->{charset} = $self->{default}->{charset};
    }

    my $message = $self->generate_message( $c, \%mime );

    if ($message) {
        my $return = sendmail( $message, { transport => $self->_mailer_obj } );

        # return is a Return::Value object, so this will stringify as the error
        # in the case of a failure.
        croak "$return" if !$return;
    }
    else {
        croak "Unable to create message";
    }
}

=item setup_attributes($c, $attr)

Merge attributes with the configured defaults. You can override this method to
return a structure to pass into L<generate_message> which subsequently
passes the return value of this method to Email::MIME->create under the
C<attributes> key.

=cut

sub setup_attributes {
    my ( $self, $c, $attrs ) = @_;

    my $default_content_type = $self->default->{content_type};
    my $default_charset      = $self->default->{charset};

    my $e_m_attrs = {};

    if (   exists $attrs->{content_type}
        && defined $attrs->{content_type}
        && $attrs->{content_type} ne '' )
    {
        $c->log->debug( 'C::V::Email uses specified content_type '
              . $attrs->{content_type}
              . '.' )
          if $c->debug;
        $e_m_attrs->{content_type} = $attrs->{content_type};
    }
    elsif ( defined $default_content_type && $default_content_type ne '' ) {
        $c->log->debug(
            "C::V::Email uses default content_type $default_content_type.")
          if $c->debug;
        $e_m_attrs->{content_type} = $default_content_type;
    }

    if (   exists $attrs->{charset}
        && defined $attrs->{charset}
        && $attrs->{charset} ne '' )
    {
        $e_m_attrs->{charset} = $attrs->{charset};
    }
    elsif ( defined $default_charset && $default_charset ne '' ) {
        $e_m_attrs->{charset} = $default_charset;
    }

    return $e_m_attrs;
}

=item generate_message($c, $attr)

Generate a message part, which should be an L<Email::MIME> object and return it.

Takes the attributes, merges with the defaults as necessary and returns a
message object.

=cut

sub generate_message {
    my ( $self, $c, $attr ) = @_;

    # setup the attributes (merge with defaultis)
    $attr->{attributes} = $self->setup_attributes( $c, $attr->{attributes} );
    Email::MIME->create( %$attr );
}

=back


=head1 TROUBLESHOOTING

As with most things computer related, things break.  Email even more so.  
Typically any errors are going to come from using SMTP as your sending method,
which means that if you are having trouble the first place to look is at
L<Email::Sender::Transport::SMTP>.  This module is just a wrapper for L<Email::Sender::Simple>,
so if you get an error on sending, it is likely from there anyway.

If you are using SMTP and have troubles sending, whether it is authentication
or a very bland "Can't send" message, make sure that you have L<Net::SMTP> and,
if applicable, L<Net::SMTP::SSL> installed.

It is very simple to check that you can connect via L<Net::SMTP>, and if you
do have sending errors the first thing to do is to write a simple script
that attempts to connect.  If it works, it is probably something in your
configuration so double check there.  If it doesn't, well, keep modifying
the script and/or your mail server configuration until it does!

=head1 SEE ALSO

=head2 L<Catalyst::View::Email::Template> - Send fancy template emails with Cat

=head2 L<Catalyst::Manual> - The Catalyst Manual

=head2 L<Catalyst::Manual::Cookbook> - The Catalyst Cookbook

=head1 AUTHORS

J. Shirley <jshirley@gmail.com>

Alexander Hartmaier <abraxxa@cpan.org>

=head1 CONTRIBUTORS

(Thanks!)

Matt S Trout

Daniel Westermann-Clark

Simon Elliott <cpan@browsing.co.uk>

Roman Filippov

Lance Brown <lance@bearcircle.net>

Devin Austin <dhoss@cpan.org>

Chris Nehren <apeiron@cpan.org>

=head1 COPYRIGHT

Copyright (c) 2007 - 2009
the Catalyst::View::Email L</AUTHORS> and L</CONTRIBUTORS>
as listed above.

=head1 LICENSE

This library is free software, you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut

1;
